src/monitor/group_state_machine.c

```diff
@@ -295,14 +341,533 @@ ProceedGroupState(AutoFailoverNode *activeNode)
+				 node->nodeId, node->nodeName, node->nodePort);
+			continue;
+		}
+
+		/*
+		 * Skip nodes that are not failover candidates (not in SECONDARY or
+		 * REPORT_LSN state).
+		 *
+		 * XXX: what about goalState instead? or in addition to?
+		 */
+		if (node->reportedState != REPLICATION_STATE_SECONDARY
+			&& node->reportedState != REPLICATION_STATE_REPORT_LSN)
+		{
+			continue;
+		}
+
+		/* count how many standby nodes have reached REPORT_LSN */
+		if (IsCurrentState(node, REPLICATION_STATE_REPORT_LSN))
+		{
+			++candidateCount;
+			++reportedLSNCount;
+		}
+		else if (node->goalState != REPLICATION_STATE_REPORT_LSN)
+		{
+			char message[BUFSIZE];
+
+			++candidateCount;
+
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state of %s:%d to report_lsn "
+				"to find the failover candidate",
+				node->nodeName, node->nodePort);
+
+			AssignGoalState(node, REPLICATION_STATE_REPORT_LSN, message);
+		}
+	}
+
+	/*
+	 * If we can failover, make sure the primary is being demoted before doing
+	 * anything else.
+	 */
+	if (candidateNode != NULL || candidateCount > 0)
+	{
+		/* shut down the primary, known unhealthy (see pre-conditions) */
+		if (IsInPrimaryState(primaryNode))
+		{
+			char message[BUFSIZE];
+
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state %s:%d to draining after it became unhealthy.",
+				primaryNode->nodeName, primaryNode->nodePort);
+
+			AssignGoalState(primaryNode, REPLICATION_STATE_DRAINING, message);
+		}
+	}
+
+	/*
+	 * If a failover is in progress, continue driving it.
+	 *
+	 * Here we have choosen a failover candidate, which is either being
+	 * promoted to being the new primary (when it already had all the most
+	 * recent WAL, or is done fetching them), or is fetching the most recent
+	 * WAL it's still missing from another standby node.
+	 */
+	if (candidateNode != NULL)
+	{
+		/*
+		 * Everyone reported their LSN, we found a candidate, and it has to
+		 * fast forward. First, the node with the most advanced LSN position
+		 * had to reach wait_cascade:
+		 */
+		if (IsCurrentState(activeNode, REPLICATION_STATE_WAIT_CASCADE) &&
+			IsCurrentState(candidateNode, REPLICATION_STATE_WAIT_FORWARD))
+		{
+			char message[BUFSIZE];
+
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state of %s:%d to fast_forward "
+				"after %s:%d converged to wait_cascade.",
+				candidateNode->nodeName, candidateNode->nodePort,
+				activeNode->nodeName, activeNode->nodePort);
+
+			AssignGoalState(candidateNode,
+							REPLICATION_STATE_FAST_FORWARD, message);
+
+			return true;
+		}
+
+		/*
+		 * The candidate had to fast forward with the activeNode, grabbing WAL
+		 * bits that only this node had. Now that's done. The activeNode has to
+		 * follow the new primary that's being promoted.
+		 */
+		if (IsCurrentState(activeNode, REPLICATION_STATE_WAIT_CASCADE) &&
+			IsCurrentState(candidateNode, REPLICATION_STATE_STOP_REPLICATION))
```

Comment:
Does only the candidate go through stop replication? I assume that doesn't cause the primary to detect a network partition. If it doesn't and wére ok with that, is the state still useful?

Reply:
Having a single node go through stop replication will not make the primary a network partition. Adding to that, we also have all the other standby nodes re-connect to the new primary when they do REPORT_LSN to SECONDARY. That happens as soon as the candidate enters STOP REPLICATION. All combined, we now expect the primary to detect the network partition (because it no longer has any steaming replica connected).

Reply:
Getting back to this point, I have now introduced a (yet another) state in between REPORT_LSN and SECONDARY. Nodes go through JOIN_SECONDARY now, and in that state we CHECKPOINT; then stop PostgreSQL. When going from JOIN_SECONDARY to SECONDARY we then edit the recovery configuration (`primary_conninfo`) and restart Postgres.

So while the selected candidate for failover goes through PREPARE_REPLICATION then STOP_REPLICATION, the other nodes go through REPORT_LSN then JOIN_SECONDARY. This gives the effect of shutting down replication streams with the primary, which will then realise it if it's alone in the loosing part of a network split.